openmls/src/group/managed_group/validation.rs

```diff
@@ -126,47 +108,78 @@ impl ManagedGroup {
            - Confirmation tag must be successfully verified
         */
 
-        // Signature verification
-        // If a signature key is provided it will be used,
-        // otherwise we take the key from the credential
-
-        let plaintext: MlsPlaintext = match signature_key {
-            Some(signature_public_key) => message
-                .plaintext
-                .verify_with_key(backend, &signature_public_key)?,
-            None => message.plaintext.verify(backend, &message.credential)?,
-        };
-
-        Ok(match plaintext.content() {
-            MlsPlaintextContentType::Application(application_message) => {
-                ProcessedMessage::ApplicationMessage(application_message.as_slice().to_vec())
+        // Add the context to the message and verify the membership tag if necessary
+        let serialized_context = self.group.context().tls_serialize_detached()?;
+
+        let context_plaintext = UnverifiedContextMessage::from_unverified_message_with_context(
+            unverified_message,
+            serialized_context,
+            self.group.epoch_secrets().membership_key(),
+            backend,
+        )?;
+
+        match context_plaintext {
+            UnverifiedContextMessage::Member(member_message) => {
+                // Signature verification
+                let verified_member_message =
+                    member_message.into_verified(backend, signature_key)?;
+
+                Ok(match verified_member_message.plaintext().content() {
+                    MlsPlaintextContentType::Application(application_message) => {
+                        ProcessedMessage::ApplicationMessage(
+                            application_message.as_slice().to_vec(),
+                        )
+                    }
+                    MlsPlaintextContentType::Proposal(_proposal) => {
```

Comment:
Would it be worth revisiting the constructors for StagedProposal and `stage_commit` to take the content instead of the whole plaintext?